spec-docs/function-types.md

```diff
@@ -0,0 +1,304 @@
+# Function Types in Kotlin on JVM
+
+## Goals
+
+* Get rid of 23 hardwired physical function classes. One of the problems with them is that they should be effectively duplicated in reflection which means a lot of physical classes in the runtime.
+* Make extension functions coercible to normal functions (with an extra parameter), so that it's possible to do `listOfStrings.map(String::length)`
+* Allow functions with more than 23 parameters, theoretically any number of parameters (in practice 255 on JVM).
+* At the same time, allow to implement Kotlin functions easily from Java: `new Function2() { ... }` and overriding `invoke` only would be the best.
+Enabling SAM conversions on Java 8 would also be terrific.
+
+## Brief solution overview
+
+* Treat extension functions almost like non-extension functions with one extra parameter, allowing to use them almost interchangeably.
+* Introduce a physical class `Function` and unlimited number of *fictitious* (synthetic) classes `Function0`, `Function1`, ... in the compiler front-end
+* On JVM, introduce `Function0`..`Function22`, which are optimized in a certain way,
+and `FunctionN` for functions with 23+ parameters.
+When passing a lambda to Kotlin from Java, one will need to implement one of these interfaces.
+* Also on JVM (under the hood) add abstract `FunctionImpl` which implements all of `Fun0`..`Fun22` and `FunN`
+(throwing exceptions), and which knows its arity.
+Kotlin lambdas are translated to subclasses of this abstract class, passing the correct arity to the super constructor.
+* Provide a way to get arity of an arbitrary `Function` object (pretty straightforward).
+* Hack `is/as Function5` on any numbered function in codegen (and probably `KClass.cast()` in reflection) to check against `Function` and its arity.
+
+## Extension functions
+
+Extension function type `T.(P) -> R` is now just a shorthand for `@kotlin.extension Function2<T, P, R>`.
+`kotlin.extension` is a **type annotation** defined in built-ins.
+So effectively functions and extension functions now have the same type,
+which means that everything which takes a function will work with an extension function and vice versa.
+
+To prevent unpleasant ambiguities, we introduce additional restrictions:
+* A value of an extension function type cannot be **called** as a function, and a value of a non-extension
+function type cannot be called as an extension. This requires an additional diagnostic which is only fired
+when a call is resolved to the `invoke` with the wrong extension-ness.
+* If an extension function **literal** argument has some shape (its parameters are written out explicitly
+in the code and it's evident that there is or there isn't a receiver parameter), this shape must exactly match
```

Comment:
If a receiver parameter type is specified in the beginning of the function literal (`{ Int.() -> }`) or in the function expression (`fun Int.() = ...`), this means that literal or expression represents an extension lambda and must only be used where the corresponding parameter's type is an extension function. The opposite is only relevant for function expressions though: if a function expression doesn't have a receiver type specified, it's a non-extension function and must only be passed to a parameter which type is a non-extension function. (If a *function literal* doesn't have a receiver type specified, it still *can* be an extension function.) Generally we want to keep as much old behavior as possible.

I'll rewrite this point in the document, currently it makes little sense